bluefin-internal/src/Bluefin/Internal/Algae/Cancellable.hs

```diff
@@ -0,0 +1,76 @@
+{-# LANGUAGE
+  BangPatterns,
+  GADTs,
+  RankNTypes,
+  ScopedTypeVariables,
+  StandaloneKindSignatures,
+  TypeOperators #-}
+{-# OPTIONS_HADDOCK not-home #-}
+
+-- | = Algebraic effects and named handlers with cancellable continuations
+module Bluefin.Internal.Algae.Cancellable
+  ( AEffect
+  , HandlerBody
+  , Handler
+  , handle
+  , call
+  , continue
+  , discontinue
+  , discontinueIO
+  ) where
+
+import Control.Exception (Exception)
+import Data.Kind (Type)
+import Bluefin.Internal (Eff, Effects, IOE, type (:>), type (:&))
+import Bluefin.Internal.DelCont
+import Bluefin.Internal.Exception.Dynamic
+import Bluefin.Internal.Algae (AEffect)
+
+-- | Interpretation of an algebraic effect @f@: a function to handle the operations of @f@
+-- with cancellable continuations.
+type HandlerBody :: Effects -> AEffect -> Effects -> Type -> Type
+type HandlerBody ex f ss a = (forall x ss0. ex :> ss0 => f x -> (Eff ss0 x -> Eff ss a) -> Eff ss a)
+
+-- | Handler to call operations of the effect @f@ with cancellable continuations.
+type Handler :: Effects -> AEffect -> Effects -> Type
+data Handler ex f s where
+  Handler :: !(PromptTag ss a s) -> HandlerBody ex f ss a -> Handler ex f s
+
+-- | Handle operations of @f@ with cancellable continuations.
+--
+-- The handle for exceptions (first argument) is only there to guide type inference.
+-- it can be either 'IOE' or 'DynExn'.
+handle ::
+  h ex ->
+  HandlerBody ex f ss a ->
+  (forall s. Handler ex f s -> Eff (s :& ss) a) ->
+  Eff ss a
+handle _ h act = reset (\p -> act (Handler p h))
+
+-- | Call an operation of @f@ with cancellable continuations.
+call :: (ex :> es, s :> es) => Handler ex f s -> f a -> Eff es a
+call (Handler p h) op = shift0 p (\k -> h op k)
+
+-- | Resume a cancellable continuation with a result.
+--
+-- In other words, this converts a cancellable continuation to a simple continuation.
+continue :: (Eff ss0 b -> Eff ss a) -> b -> Eff ss a
+continue k = k . pure
+
+-- | Cancel a continuation: resume by throwing a (dynamic) exception.
+discontinue :: (Exception e, ex :> es0) => DynExn ex -> (Eff es0 b -> Eff es a) -> e -> Eff es a
+discontinue ex k e = k (throw ex e)
```

Comment:
I don't understand what role `Exception` plays here (and in `discontinueIO`). Could we not use a Bluefin (i.e. well-scoped) exception instead?

Reply:
In fact I see native `Exception` used in other places too. Is that necessary? One of the things I most value about Bluefin is that *any* type can be thrown as an exception.

Reply:
It would work with bluefin exceptions but only because they are partly observable as they rely on native exceptions: you can `catch` them as `SomeException`.

With a truly scoped exception, the continuation would just not see it: `k (throw ex e)` would be equivalent to `throw ex e`, doing nothing useful in `k`. With "real world" resources, you want to be ready to clean up on any error, so dynamic exceptions seems an inherently better fit.

Reply:
I guess ultimately what I'm asking is, can we get rid of that `Exception` constraint on `e`? Perhaps by putting `e` as a type argument to `DynExn`?

Reply:
Oh, yes that seems feasible.